core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1972,62 +1972,62 @@ protected Type visitExtract(Extract node, StackableAstVisitorContext<Context> co
-                    }
-                    break;
-                case TIMEZONE_MINUTE:
-                case TIMEZONE_HOUR:
-                    if (!(type instanceof TimestampWithTimeZoneType) && !(type instanceof TimeWithTimeZoneType)) {
-                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
-                    }
-                    break;
-                default:
-                    throw new UnsupportedOperationException("Unknown field: " + field);
+            if (Extract.Field.YEAR.equals(field)) {
+                checkExtractMonth(node, type, field);
+                outputType = INTEGER;
+            }
+            else if (Extract.Field.MONTH.equals(field)) {
+                checkExtractMonth(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.DAY.equals(field)) {
+                if (!(type instanceof DateType) &&
+                        !(type instanceof TimestampType) &&
+                        !(type instanceof TimestampWithTimeZoneType) &&
+                        !(type.equals(INTERVAL_DAY_TIME))) {
+                    throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
+                }
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.QUARTER.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.WEEK.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.DAY_OF_MONTH.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.DAY_OF_WEEK.equals(field) || Extract.Field.DOW.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.DAY_OF_YEAR.equals(field) || Extract.Field.DOY.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = SMALLINT;
+            }
+            else if (Extract.Field.YEAR_OF_WEEK.equals(field) || Extract.Field.YOW.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
+                outputType = INTEGER;
+            }
+            else if (Extract.Field.HOUR.equals(field) || Extract.Field.MINUTE.equals(field) || Extract.Field.SECOND.equals(field)) {
+                checkExtractSecond(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.TIMEZONE_HOUR.equals(field) || Extract.Field.TIMEZONE_MINUTE.equals(field)) {
+                checkExtractTimezoneHour(node, type, field);
```

Comment:
This refactoring seems unnecessary. All that seems to be needed is to add outputType to each branch of the switch and possibly splitting some of them if they need to produce different types.

Reply:
Yes, you are right. While reading the code again, I think that setting `outputType` should be enough

Reply:
I gave again a thought to this topic (the same as I did a while back while I initially implemented the change).

I find the usage of the `switch` inappropriate in the context of introducing the `outputType` logic.
Due to the fact that the `switch` statement is used rather seldom, it is rather misleading because it executes all the expressions that follow the matching expression (until finding a `break` statement).

I opted to switch the code to `if` in order to avoid checking twice what kind of `field` are we dealing with:

```
    switch(field):
      ....
      case QUARTER:
                case WEEK:
                case DAY_OF_MONTH:
                case DAY_OF_WEEK:
                case DOW:
                case DAY_OF_YEAR:
                case DOY:
                case YEAR_OF_WEEK:
                case YOW:
                    if (!(type instanceof DateType) &&
                            !(type instanceof TimestampType) &&
                            !(type instanceof TimestampWithTimeZoneType)) {
                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
                    }
                    if (Extract.Field.WEEK.equals(field)){
                        outputType = TINYINT;
                   } else if (EXTRACT.Field.DAY_OF_MONTH.equals(field)) {
                       outputType = TINYINT;
                   } ...
                   ...
``` 

This is why I find the switch to the boring `if` - `else if` statements more appropriate in the current context.